engine/src/main/java/org/camunda/bpm/engine/impl/persistence/entity/TimerEntity.java

```diff
@@ -100,6 +110,59 @@ protected void preExecute(CommandContext commandContext) {
     }
   }
 
+  protected String parseExpression(CommandContext commandContext) {
+    String expressionValue = null;
+    String expression = jobDefinition.getJobConfiguration().substring(7);
+    try {
+      expressionValue = commandContext.getProcessEngineConfiguration()
+          .getExpressionManager()
+          .createExpression(expression)
+          .getValue(execution).toString();
+    } catch (Exception e) {
+      throw ProcessEngineLogger.UTIL_LOGGER.exceptionWhileParsingCycleExpresison(expression, e);
+    }
+    return expressionValue;
+  }
+
+  protected boolean isReevaluateTimeCycleWhenDue(CommandContext commandContext) {
+    return commandContext.getProcessEngineConfiguration().isReevaluateTimeCycleWhenDue();
+  }
+
+  protected boolean isCycleExpression() {
+    // Note timer cycle configuration is constructed in BpmnParse#parseTimer
+    String jobConfiguration = jobDefinition.getJobConfiguration();
+    return jobConfiguration.contains("CYCLE: #") || jobConfiguration.contains("CYCLE: $");
+  }
+
+  protected String adjustRepeatBasedOnNewExpression(String expressionValue) {
+    String changedRepeat;
+    SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss");
+    if (expressionValue.startsWith("R")) { // changed to a repeatable interval
+      if (repeat.startsWith("R") ) {
+        if (isSameRepeatCycle(expressionValue)) {
+          // the same repeatable interval => keep the start date
+          changedRepeat = repeat;
+        } else {
+          // different repeatable interval => change the start date
+          changedRepeat = expressionValue.replace("/", "/" + sdf.format(ClockUtil.getCurrentTime()) + "/");
+        }
+      } else {
+        // was a cron expression => change the start date
+        changedRepeat = expressionValue.replace("/", "/" + sdf.format(ClockUtil.getCurrentTime()) + "/");
+      }
+    } else {
+      // changed to a cron expression
+      changedRepeat = expressionValue;
+    }
+    return changedRepeat;
+  }
+
+  protected boolean isSameRepeatCycle(String expressionValue) {
+    String[] currentRepeat = repeat.split("/");      // "R3/date/PT2H"
+    String[] newRepeat = expressionValue.split("/"); // "R3/PT2H"
```

Comment:
❓ Since we explicitly check for `length==2` in https://github.com/camunda/camunda-bpm-platform/blob/e573859c094bff126b86cffc2d0295068a7d8ad1/engine/src/main/java/org/camunda/bpm/engine/impl/jobexecutor/TimerDeclarationImpl.java#L161-L167 of the split, should we consider this here as well? What if the repeat value the expression yields is "corrupt"?

Reply:
I will add a check and consider adding a test case too.